onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h

```diff
@@ -0,0 +1,258 @@
+name was chosen to match with the dnnl name when running in inference mode. The
+dst for destination matches the naming scheme for dnnl in inference mode.  When
+in training mode primitive_src would make more since but we are reusing the
+already provided output parameter.
+
+The dnnl backward primitive description requires the forward primitive
+description. Since forward primitive description requires the memory
+description (md) from the forward Relu operator. Since the actual memory
+description from forward Relu is not avalible we are using the memory
+description from the x input. With the belive that the important parts of
+the memory descrption are the shape and dimentions which match the x input.
+*/
+
+namespace onnxruntime {
+namespace ort_dnnl {
+template <typename T>
+class DnnlReluGrad : public DnnlKernel {
+ public:
+  DnnlReluGrad(const DnnlNode& node,
+               DNNLExecutionProvider* provider,
+               const Provider_NodeAttributes& attributes,
+               const std::string attributes_prefix = "") : DnnlKernel(node, provider) {
+    ORT_UNUSED_PARAMETER(attributes);
+    ORT_UNUSED_PARAMETER(attributes_prefix);
+  }
+
+  void AddForwardDnnlKernel(std::shared_ptr<DnnlRelu<T>> relu_fwd) {
+    relu_fwd_ = relu_fwd;
+  }
+
+  void CreatePrimitives(const OrtCustomOpApi* api,
+                        OrtKernelContext* context,
+                        const std::unordered_map<dnnl::engine::kind, dnnl::engine>& dnnl_engine,
+                        std::vector<dnnl::primitive>& net,
+                        std::vector<std::unordered_map<int, dnnl::memory>>& net_args) override {
+    dnnl::engine cpu_engine;
+    dnnl::engine engine_to_use;
+    std::unordered_map<dnnl::engine::kind, dnnl::engine>::const_iterator iter = dnnl_engine.find(dnnl::engine::kind::cpu);
+    if (iter != dnnl_engine.end()) {
+      cpu_engine = (dnnl::engine)iter->second;
+      engine_to_use = cpu_engine;
+    }
+    #if 0 // TODO update relugrad for gpu
+    gpu_available_ = false;
+    dnnl::engine gpu_engine;
+    iter = dnnl_engine.find(dnnl::engine::kind::gpu);
+    if (iter != dnnl_engine.end()) {
+      gpu_engine = (dnnl::engine)(iter->second);
+      gpu_available_ = true;
+      engine_to_use = gpu_engine;
+      LOGS_DEFAULT(INFO) << "gpu engine found" << std::endl;
+    }
+    #endif
+    if (!relu_fwd_){
+      ORT_THROW("Unable to find forward pass Relu opt. "
+                "Verify AddForwardDnnlKernel was called after DnnlReluGrad was created.");
+    }
+
+    Ort::CustomOpApi ort{*api};
+    int input_index = mklnode_ptr_->input_start_index < 0 ? 0 : mklnode_ptr_->input_start_index;
+
+    TensorShape xgrad_shape;
+    TensorShape x_shape;
+    if (mklnode_ptr_->parent_nodes.empty()) {
+      // convert the onnx gradient Tensor to dnnl::memory and dnnl::memory::desc
+      const OrtValue* xgrad_input_tensor = ort.KernelContext_GetInput(context, input_index);
```

Comment:
At the start of the session, the `CreatePrimitives` would be called once for each ReluGrad node in the graph. This will set up the input/output tensors based on the input graph. This only sets up the node it is not executed at this stage. There is an instance of the DnnlReluGrad for each ReluGrad node. The should be no synchronization problems between multiple threads. 

 When the ReluGrad node is executed it will call the `Bind()` function to get the memory from OnnxRuntime then the actual operator is run. (See the `SubgraphPrimitive::Compute(...)` found in dnnl_func_kernel.cc)

Reply:
Thanks for bearing with me! This EP is new territory for me, so I took a look at dnnl_func_kernel.cc to understand. It looks like the thread-local maps in `PrimitivePool<T>::GetMap()` are taking care of the thread-safety issue.